### PR TITLE
Add a no-op Sync method to satisfy latest datastore interface

### PR DESCRIPTION
**Pull request by kishansagathiya**

### CODE REVIEW

File: crdt.go [hunk 1]
```diff
@@ -598,6 +598,15 @@ func (store *Datastore) Delete(key ds.Key) error {
 	return store.publish(delta)
 }
 
+// Sync is a no-op.
+// Before broadcasting deltas we run processNode making sure that every
+// operation (including Put and Delete) applied to this store take effect
+// (delta is merged) before returning.
+func (store *Datastore) Sync(key ds.Key) error {
+	store.logger.Warning("sync is a no-op")
```
Comment (Stebalien) [inline crdt.go]: If sync doesn't do anything, there's no need to log. We should just return nil and move on.

**Comment by aschmahmann** [comment]
> For the moment we are not upgrading to the new datastore version, until this is all figured out.

@hsanjuan 👍 that makes a lot of sense. I notice that go-ds-crdt uses `DAGService`s. In go-ipfs we worked around not changing the `Blockstore`, `Blockservice` and `DagService` interfaces to accommodate asynchronous writes, however if it's useful we may do so going forward.

If while figuring out where to Sync in go-ds-crdt you realize that we really need to update some other interfaces feel free to tag me and we can come up with a plan.

**Comment by Stebalien** [comment]
> If Sync should be a no-op then we should be calling store.Sync() at a bunch of places to make sure that the underlying (potentially async datastore) is synced after every write... otherwise we need to keep tabs on what internal keys are related to the external keys as provided by the user and sync those, otherwise we need to put big warnings that the underlying datastore should not be async or data might be lost in potentially very bad ways.

The datastores are still sync by default, we're just:

* Adding an _option_ to the datastores to disable synchronous writes (well, this option already exists).
* Adding a function (Sync) to explicitly sync everything under a given key.

Updating the datastores and never calling sync anywhere is perfectly safe unless you also _explicitly_ disable synchronous writes.

**Comment by hsanjuan** [comment]
> Updating the datastores and never calling sync anywhere is perfectly safe unless you also _explicitly_ disable synchronous writes.

Yes, that's clear. The issue here is that the user provides an abstract underlying `Datastore` implementation which may be, at the bottom of all wrappers, be configured as sync or not, and this library has no way of knowing, so we have to assume it *may* be async... 

I've read the discussion, but the way this has been added is problematic for cases like this. Other things like a Batching Datastore or a PersistentDatastore can be checked for, but now all datastores are async by the mere fact that they are forced to provide the Sync method, regardless of whether it's a no-op or not.

**Comment by Stebalien** [comment]
> Yes, that's clear. The issue here is that the user provides an abstract underlying Datastore implementation which may be, at the bottom of all wrappers, be configured as sync or not, and this library has no way of knowing, so we have to assume it may be async...

Ah, now I see. I saw "sync is a no-op" and assumed that sync was _actually_ a no-op. It isn't a no-op, it's just not implemented.

Why not implement it? At a minimum, just calling `Sync("/")` on the underlying datastore should work (and should be a no-op on datastores with sync writes enabled).

**Comment by hsanjuan** [comment]
> Why not implement it? At a minimum, just calling `Sync("/")` on the underlying datastore should work (and should be a no-op on datastores with sync writes enabled).

Right, that can be done, or we can actually see what is the way to actually just sync relevant subkeys to the given key in the underlying datastore. I never argued for not implementing. I'm saying it should not be a no-op.

**Comment by hsanjuan** [comment]
Overriden by #22.

**Pull request closed without merging**